### PR TITLE
Tests: Add simple tests for Android 2.3

### PR DESCRIPTION
**Pull request by mzgol**
Fixes gh-2505
Refs gh-2483

@jquery/core This is an initial stab at simple tests for Android 2.3 but please review what I have before I write tests for other modules.

I've created a separate "module" called `basic` to contain those selected tests; I also added it to the list run by regular browsers so that we immediately know if those tests are ok.

Note that tests are *very* basic but testing the whole API even without edge cases (however you define them) would be way too much; if we run more than 10% of assertions in Android 2.3 it will be unstable (older browsers fare better with many modules with few tests in them rather than one module with all). I'd rather remove some CSS tests from this list than add more.

I've temporarily modified our Jenkins jobs to see how that would look:

1. the regular periodic run without Android 2.3: http://swarm.jquery.org/job/1009
2. the basic run just for Android 2.3: http://swarm.jquery.org/job/1011

Even these 18 CSS tests seem to *sometimes* run very slow on Android 2.3: they were running for 11 minutes & timed out, after restarting they passed almost immediately...

### CODE REVIEW

**Comment by dmethvin** [comment]
Anything that goes through only `basic` support doesn't test many code paths at all, yet most of them work with Android 2.3 don't they? It seems like we'd want to instead specifically exclude (and document) the cases we know don't work, like `reliableMarginRight`. That list of tests could be pulled out into its own module perhaps?

**Comment by mzgol** [comment]
> Anything that goes through only `basic` support doesn't test many code paths at all, yet most of them work with Android 2.3 don't they?

Most currently do. The fact that `basic` won't test many "basic" things is one of the reasons why I'd prefer to just drop Android 2.3 on `master` and leave it just on `compat` to avoid such weird compatibility state but only @markelog supported me in this view and we decided we'll be just running basic tests, i.e. what this PR tries to do.

But yeah, I'm still not terribly convinced.

> It seems like we'd want to instead specifically exclude (and document) the cases we know don't work, like reliableMarginRight. That list of tests could be pulled out into its own module perhaps?

Pulling such tests to a separate module would be catering to Android 2.3 too much, IMO; we'd break module separation just because of one browser. IMO it's easier (& better) to just skip problematic tests as we started to do now.

The problem is not only failing tests, though. Those tests run extremely slow, usually ~1/3 of the modules time out and I have to restart those runs a few times before they stabilize and (mostly) pass to be sure we didn't regress. That's quite tedious. Running just very basic tests should at least help with this problem.

File: test/unit/basic.js [hunk 1]
```diff
@@ -0,0 +1,84 @@
+module( "basic", {
+	teardown: moduleTeardown
+} );
+
+if ( jQuery.css ) {
+	test( "css", function() {
+		expect( 18 );
```
Comment (gibson042) [inline test/unit/basic.js]: This already looks like too much... I'd expect to see one `test` per module (as you have), but no more than 10 assertions per test, and hopefully even fewer. For the CSS module, maybe one assertion like `assert.equal( div.css( "width", "50px" ).css( "width" ), "50px", ".css getter/setter" )` and one for `.show`/`.hide` and call it done.

**Comment by mzgol** [comment]
PR updated (traversing & wrap tests to be written).

**Comment by mzgol** [comment]
Ready for final review.

**Comment by dmethvin** [comment]
Now I'm really wanting to get some code coverage stats. :)

File: test/unit/basic.js [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+	teardown: moduleTeardown
+} );
+
+if ( jQuery.ajax ) {
+test( "ajax", function( assert ) {
+	expect( 4 );
+
+	var requestsLeft = 3;
+
+	stop();
```
Comment (gibson042) [inline test/unit/basic.js]: Could you update these to use `assert.async` instead of `stop`/`start`?

Comment (markelog) [inline test/unit/basic.js]: We can do this across the all unit files, so it could be done later.

File: test/unit/basic.js [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+	assert.ok( jQuery.isWindow( window ), "jQuery.isWindow(window)" );
+	assert.ok( !jQuery.isWindow( 2 ), "jQuery.isWindow(Number)" );
+
+	assert.strictEqual( jQuery.inArray( 3, [ "a", 6, false, 3, {} ] ), 3, "jQuery.inArray - true" );
+	assert.strictEqual( jQuery.inArray( 3, [ "a", 6, false, "3", {} ] ), -1, "jQuery.inArray - false" );
+
+	assert.strictEqual( elem.get( 1 ), elem[ 1 ], ".get" );
+	assert.strictEqual( elem.first()[ 0 ], elem.get( 0 ), ".first" );
+	assert.strictEqual( elem.last()[ 0 ], elem.get( 1 ), ".last" );
```
Comment (gibson042) [inline test/unit/basic.js]: Let's define expected values without `.get`.

**Comment by mzgol** [comment]
> Now I'm really wanting to get some code coverage stats. :)

I had significant problems with getting coverage to be generated in our current setup (most likely I did everything wrong :D), in the end I hacked it around in PhantomJS but excluding jQuery.ajax tests as I had problems with getting it running and I got:

* Statements: 42.14% (1369 / 3249)
* Branches: 28.39% (926 / 3262)
* Functions: 42.58% (244 / 573)
* Lines: 42.14% (1369 / 3249)

File: test/unit/basic.js (hunk 2 above)
Comment (mzgol) [inline test/unit/basic.js]: Sure, since I use the new syntax in these files I may as well get it right.

File: test/unit/basic.js (hunk 3 above)
Comment (mzgol) [inline test/unit/basic.js]: Done.

**Comment by markelog** [comment]
These stats for all modules? Could run just `basic` ones?

**Comment by mzgol** [comment]
No, just for `basic`.

**Comment by mzgol** [comment]
Travis fails because of problems with installing old jsdom... Can we get #2526 landed ASAP?

**Comment by gibson042** [comment]
This looks nice. :+1:

**Comment by timmywil** [comment]
:+1:

File: test/unit/basic.js [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+		requestsLeft = 3;
+
+	jQuery.ajax({
+		type: "GET",
+		url: url( "data/name.php?name=foo" ),
+		success: function( msg ) {
+			assert.strictEqual( msg, "bar", "Check for GET" );
+			requestsLeft--;
+			if ( !requestsLeft ) {
+				done();
```
Comment (gibson042) [inline test/unit/basic.js]: You don't need to do the tracking yourself; just use `assert.async` multiple times (cf. [test/unit/deferred.js](https://github.com/jquery/jquery/blob/1c59b308d201d6dd0f0aed2ad0256d01b9f68047/test/unit/deferred.js#L411-L421)).

Comment (mzgol) [inline test/unit/basic.js]: Updated, although that doesn't save that much. Ideally `assert.async()` would accept a parameter indicating how many async operations are expected and that would require calling `done()` this many times.

BTW, `done.pop().call()` is not needed, it's just `done.pop()()` after all.

Comment (gibson042) [inline test/unit/basic.js]: Of course you're right about `.call()`, and I agree on the former point. In fact, I meant to create a ticket for that earlier, but it fell off my radar. Anyway, corrected now: jquery/qunit/issues/843

Comment (mzgol) [inline test/unit/basic.js]: Thanks for creating a ticket, I should've done it myself. :)
